guides/strategies-that-shrink.rst

```diff
@@ -0,0 +1,213 @@
+need a better trick. Fortunately we have one which works *at a low level*;
+we use it in several crucial places internally but it may not be of any use
+for external strategies.
+
+Aside from anything else, **this uses private internals which may be broken
+in any patch release**.  If you intend to try this, please contact us for
+advice first and we may be able to provide a more robust way to do it.
+
+Basically, we ensure that the example we draw has a simple representation
+in the buffer that we can shrink from, even if that's not how we really
+generated it.
+
+First, the "one-shot filter":
+
+- Try to draw an example from ``Maybe``, a strategy that might or might
+  not give us a valid example on the first attempt.  If this succeeds,
+  we're done.
+- If it fails, use an internal API to mark it as invalid.  Then, create
+  a more expensive or less consistent (see below) strategy that will
+  generate an example which *could have come from* ``Maybe`` and draw
+  that example.  Mark this invalid too.  Finally, calculate the buffer
+  that would generate this example from ``Maybe``, and use another internal
+  API to append it to the buffer.
+
+When we go to shrink such a buffer, there are two possibilities:  either
+the first attempt to draw from ``Maybe`` worked, and it shrinks the easy
+way.  Otherwise, we delete the two draws that were marked invalid, and
+try to draw an example from ``Maybe`` using the bytes we wrote - and if
+we got this right, that works and gives the same example we got during
+the generation phase!
+
+(you might be able to guess we use this technique as little as possible)
+
+The second variant lets us use a simple strategy in the shrinking phase,
+but generate (most of) our new examples from a more complex strategy with
+a different distribution.  For example:
+
+1. Draw a byte, a unicode category, and a character from that category.
+2. If the byte is nonzero, write the index of that character to the buffer.
+   Otherwise, draw an index and use the character at that index instead.
+
+This experimental approach generates exotic `unicode characters
+<https://github.com/Zac-HD/hypothesis/blob/f1f951d67f9161947a298db8d5aa12b24a633c2b/hypothesis-python/src/hypothesis/searchstrategy/strings.py#L78-L97>`_
+much more often than uniform generation, but preserves our codepoint-based
+shrinking order.  `See here for more <https://github.com/HypothesisWorks/hypothesis/issues/1401>`_.
+
+
+Consistent representation of repeated choices
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This comes up when choosing which action to take in stateful testing, where
+we would like deleting any single step to still give us a valid buffer (though
+perhaps one which doesn't reproduce the bug).
+
+That means that we need to choose from the same list of options every time,
+even though some might be ruled out by preconditions.  We now implement this
+mainly via the 'shrink open' trick (it's very versatile), but have used other
```

Comment:
I was just looking at this code and I don't think we do. The way we implement it is in https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/stateful.py#L631 and the logic is:

1. Draw an integer `i` unconditionally.
2. If the rule at `i`'s precondition is satisfied, return that.
3. Filter the list of rules and draw a rule from that.
4. Write the rule index of the drawn rule to the data stream.

As a result the `"XX"` block program will shrink this to the right thing by deleting the first two draws, but that's not quite the same as the shrink open drink I think.

Reply:
I think that's the "one-shot filter", just without bothering to mark the earlier draws as invalid.

Reply:
Hmm. I think I need to reread that section more carefully then, because I think there are some important differences, but I'll have to think about what they are a bit more precisely.

Reply:
No, you're right that there's a distinction... just hard to put my finger on it.

This process of classifying and naming tactics might be the most useful part of the whole writeup - "the limits of my (natural) language are the limits of my program", if you like.